### PR TITLE
Fix the exception behavior when `config.string` contains ${VAR} in the comments.

### LINKED ISSUE #16008: Commented out variable substitutions may still be evaluated in pipelines.yml
1. Logstash version (e.g. `bin/logstash --version`)
Latest versions of `7.17`, `8.12` and `main`
2. Logstash installation source (e.g. built from source, with a package manager: DEB/RPM, expanded from tar or zip archive, docker)
Built from source, tar.gz and Logstash on ECK (docker image)

**JVM** (e.g. `java -version`):
Bundled JVM

**Description of the problem including expected versus actual behavior**:
When running a pipeline defined in `pipelines.yml`, pipelines defined in `config.string` that have commented out lines, will have any variable substitution (ENV or keystore) present in those commented out lines, evaluated.

This was discovered during ECK testing when commenting out a plugin definition that contained a non-existent ENV variable normally provided by `ElasticsearchRef`

**Steps to reproduce**:

Create a `pipelines.yml` that looks like:

```
- pipeline.id: test
  config.string: |
    input { generator {} } 
    output { 
      nil {}
      # ${I_AM_NOT_HERE}
    }   
```
and run `bin/logstash`.

Logstash will fail to start (regardless of whether there are also valid pipelines specified), giving the following error:

```
[2024-03-14T14:15:16,947][ERROR][logstash.config.sourceloader] Could not fetch all the sources {:exception=>LogStash::ConfigurationError, :message=>"Cannot evaluate `${I_AM_NOT_HERE}`. Replacement variable `I_AM_NOT_HERE` is not defined in a Logstash secret store or as an Environment entry and there is no default value given.", :backtrace=>["/Users/robbavey/code/logstash/logstash-core/lib/logstash/util/substitution_variables.rb:75:in `block in replace_placeholders'", "org/jruby/RubyString.java:3186:in `gsub'", "/Users/robbavey/code/logstash/logstash-core/lib/logstash/util/substitution_variables.rb:60:in `replace_placeholders'", "/Users/robbavey/code/logstash/logstash-core/lib/logstash/util/substitution_variables.rb:44:in `deep_replace'", "/Users/robbavey/code/logstash/logstash-core/lib/logstash/util/substitution_variables.rb:35:in `block in deep_replace'", "org/jruby/RubyHash.java:1610:in `each'", "/Users/robbavey/code/logstash/logstash-core/lib/logstash/util/substitution_variables.rb:34:in `deep_replace'", "/Users/robbavey/code/logstash/logstash-core/lib/logstash/util/substitution_variables.rb:41:in `block in deep_replace'", "org/jruby/RubyArray.java:1983:in `each'", "org/jruby/RubyEnumerable.java:1212:in `each_with_index'", "/Users/robbavey/code/logstash/logstash-core/lib/logstash/util/substitution_variables.rb:40:in `deep_replace'", "/Users/robbavey/code/logstash/logstash-core/lib/logstash/config/source/multi_local.rb:33:in `pipeline_configs'", "/Users/robbavey/code/logstash/logstash-core/lib/logstash/config/source_loader.rb:76:in `block in fetch'", "org/jruby/RubyArray.java:2800:in `collect'", "/Users/robbavey/code/logstash/logstash-core/lib/logstash/config/source_loader.rb:75:in `fetch'", "/Users/robbavey/code/logstash/logstash-core/lib/logstash/agent.rb:179:in `converge_state_and_update'", "/Users/robbavey/code/logstash/logstash-core/lib/logstash/agent.rb:117:in `execute'", "/Users/robbavey/code/logstash/logstash-core/lib/logstash/runner.rb:431:in `block in execute'", "/Users/robbavey/code/logstash/vendor/bundle/jruby/3.1.0/gems/stud-0.0.23/lib/stud/task.rb:24:in `block in initialize'"]}
[2024-03-14T14:15:16,949][ERROR][logstash.agent           ] An exception happened when converging configuration {:exception=>RuntimeError, :message=>"Could not fetch the configuration, message: Cannot evaluate `${I_AM_NOT_HERE}`. Replacement variable `I_AM_NOT_HERE` is not defined in a Logstash secret store or as an Environment entry and there is no default value given."}
```

Note that this does not happen when pipelines are configured in config files, either directly via `-f`, or referred to in `pipelines.yml`

### PR DESCRIPTION
**Pull request by mashhurs**
## Release notes

[rn:skip]

## What does this PR do?
We were applying var substitution once we load from YAML and it affects when we use `config.string`. Other options like `path.config` since points to path, the configs will be [loaded later](https://github.com/mashhurs/logstash/blob/main/logstash-core/lib/logstash/config/source/local.rb#L202-L216) and LSCL rules will be processed when compiling the pipeline. This is _why_ we are experiencing the issue with `config.string` _only_.

This PR standardizes (same sequence will be with `config.string` as well) the way we are loading configs and compiling.

## Why is it important/What is the impact to the user?
Probability might be low but it is still not an expectation to resolve the `${VAR}` if commented out. So, users may have a better experience with this change.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- ~~[ ] I have made corresponding changes to the documentation~~
- ~~[ ] I have made corresponding change to the default configuration files (and/or docker env variables)~~
- [x] I have added tests that prove my fix is effective or that my feature works

## Author's Checklist

- [ ]

## How to test this PR locally

- pull the change
- use following config in the `pipelines.yml`
  ```
   - pipeline.id: test
  config.string: |
    input { generator {} } #${I_AM_NOT_HERE}
    filter { sleep { time => 100 } }
    output {
    #${I_AM_NOT_HERE}
      stdout {
        codec => rubydebug
      }
      # ${I_AM_NOT_HERE}
    }
  ```
- Run Logstash with `bin/logstash --debug --config.debug` to see the what config was applied. 
- Logstash runs without crash. Before the change, Logstash was failing `Cannot evaluate ${I_AM_NOT_HERE}... error`
  - and config will be applied after filtering out `${I_AM_NOT_HERE}` with the comments
    ```
     ----------------------P[input-generator{"count"=>5}|[string]config_string:1:9:```
     generator { count => 5 }
     ```] -> __QUEUE__
     __QUEUE__ -> filter_to_output
     filter_to_output -> P[output-stdout{"codec"=>#<Java::OrgLogstashConfigIrImperative::PluginStatement:0x3344b612>}|     
     [string]config_string:3:1:```
     stdout {
     codec => rubydebug 
     }
     ```]
    ```

## Related issues

- Closes #16008

### CODE REVIEW

File: logstash-core/lib/logstash/util/substitution_variables.rb [hunk 1]
```diff
@@ -36,7 +38,6 @@ def deep_replace(value)
       end
     else
       if value.is_a?(Array)
-        value_array_index = 0
```
Comment (mashhurs) [inline logstash-core/lib/logstash/util/substitution_variables.rb:39]: review note: this isn't used anywhere.

File: logstash-core/spec/logstash/util/substitution_variables_spec.rb [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+        }'
+      expect(substitution_var_test_class.replace_placeholders(config_with_commented_var_line))
+        .to eq("tcp {\nport => 12345 \n}")
+    end
+
+  end
+
+  describe "exclude config comments" do
+
+    it "doesn't exclude comments if config string dosn't contain ${VAR}" do
```
Comment (mashhurs) [inline logstash-core/spec/logstash/util/substitution_variables_spec.rb:40]:
```suggestion
    it "uses as it is if config string doesn't contain ${VAR}" do
```

File: logstash-core/spec/logstash/util/substitution_variables_spec.rb [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+      single_line_config = "{ sleep { time => 1 } }"
+      multiline_config = "stdout {
+        codec => rubydebug
+      }"
+      expect(substitution_var_test_class.exclude_config_comments(single_line_config)).to eq(single_line_config)
+      expect(substitution_var_test_class.exclude_config_comments(multiline_config)).to eq(multiline_config)
+    end
+
+    it "removes the line value if commented" do
```
Comment (mashhurs) [inline logstash-core/spec/logstash/util/substitution_variables_spec.rb:49]:
```suggestion
    it "removes the config line if commented" do
```

File: logstash-core/spec/logstash/util/substitution_variables_spec.rb [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+      config_with_commented_var_line = 'tcp {
+          port => 12345
+          #"${TCP_PORT}"
+        }'
+      expect(substitution_var_test_class.exclude_config_comments(config_with_commented_var_line))
+        .to eq("tcp {\nport => 12345\n}")
+    end
+
+    it "wipes out comments if config string contains ${VAR}" do
```
Comment (mashhurs) [inline logstash-core/spec/logstash/util/substitution_variables_spec.rb:58]:
```suggestion
    it "wipes out comments if ${VAR} is in the comment of the config string" do
```

**Review by yaauie — COMMENTED** [review]
While this solves the _specific_ use-case of `config.string` having a syntax that hides comments between a `#` char and a newline, it effectively does so by infecting the substitution variables handler with knowledfe about the syntax of _one of_ the things it parses.

We _may_ be better served to run the substitution variables on a key-by-key basis for settings, allowing settings like `config.string` to opt _out_ of preemptive replacements because we know that they will be replaced later after the value has been syntactically parsed. This will ensure that the replacements run on `config.string` happen in a single pass, and not piecemeal in two passes.

File: logstash-core/src/main/java/org/logstash/config/ir/PipelineConfig.java [hunk 5]
```diff
@@ -172,7 +172,6 @@ public void displayDebugInformation() {
             logger.debug("\n\n{}", configPart.getText());
         }
         logger.debug("Merged config");
-        logger.debug("\n\n{}", this.configString());
```
Comment (mashhurs) [inline logstash-core/src/main/java/org/logstash/config/ir/PipelineConfig.java:175]: review note: this line makes twice same config print (when `--debug` and `--config.debug` enabled) which makes us confusing..., removing.

File: logstash-core/lib/logstash/config/source/multi_local.rb [hunk 6]
```diff
@@ -30,7 +30,7 @@ def initialize(settings)
     end
 
     def pipeline_configs
-      pipelines = deep_replace(retrieve_yaml_pipelines())
+      pipelines = retrieve_yaml_pipelines
```
Comment (mashhurs) [inline logstash-core/lib/logstash/config/source/multi_local.rb:33]: okay, things are clear now.
We do not _really_ need var substitution when loading the YAML. Once we have pipeline definitions from YAML, we  [properly load the configs](https://github.com/mashhurs/logstash/blob/main/logstash-core/lib/logstash/config/source/local.rb#L202-L216) based on the given options (`config.string`, `config.path`). Then, imperative pipeline compile operation executes [the LSCL](https://github.com/mashhurs/logstash/blob/b0948a78c01f5abc6183f9b37f037b50afe63cad/logstash-core/src/main/java/org/logstash/config/ir/ConfigCompiler.java#L94) which wipes out the comments based on the [rule](https://github.com/mashhurs/logstash/blob/7dde084ba66ea0572da26a6a673905202ec84ebb/logstash-core/lib/logstash/compiler/lscl/lscl_grammar.treetop#L9-L11)

**Comment by mashhurs** [comment]
https://www.elastic.co/guide/en/cloud-on-k8s/master/k8s-logstash-configuration.html

> While this solves the _specific_ use-case of `config.string` having a syntax that hides comments between a `#` char and a newline (or EOF), it effectively does so by infecting the substitution variables handler with knowledge about the syntax of _one of_ the things it parses.
> 
> We _may_ be better served to run the substitution variables on a key-by-key basis for settings, allowing settings like `config.string` to opt _out_ of preemptive replacements because we know that they will be replaced later after the value has been syntactically parsed. This will ensure that the replacements run on `config.string` happen in a single pass, and not piecemeal in two passes.

Ha.. what a fun! Your point totally makes sense (which I didn't know), thank you. 
Now, things are clear that we do not necessarily need to apply var substitution when loading the YAML. Later on, we [properly load the configs](https://github.com/mashhurs/logstash/blob/main/logstash-core/lib/logstash/config/source/local.rb#L202-L216) based on the given options (config.string, config.path). Then, imperative pipeline compile operation executes [the LSCL](https://github.com/mashhurs/logstash/blob/b0948a78c01f5abc6183f9b37f037b50afe63cad/logstash-core/src/main/java/org/logstash/config/ir/ConfigCompiler.java#L94) which wipes out the comments based on the [rule](https://github.com/mashhurs/logstash/blob/7dde084ba66ea0572da26a6a673905202ec84ebb/logstash-core/lib/logstash/compiler/lscl/lscl_grammar.treetop#L9-L11). Let me please know if I am still missing something.

File: logstash-core/lib/logstash/util/substitution_variables.rb (hunk 1 above)
Comment (andsel) [reply logstash-core/lib/logstash/util/substitution_variables.rb:39]: :+1 it's a leftover from #13603

File: logstash-core/src/main/java/org/logstash/config/ir/PipelineConfig.java (hunk 5 above)
Comment (andsel) [reply logstash-core/src/main/java/org/logstash/config/ir/PipelineConfig.java:175]: It's duplicated only when the config in composed by just one SourceWithMetadata, but when the config is loaded from multiple files, like when glob operator is used, it logs, every single section, and then also the effective pipeline definition as it's passed down to the compiler.
So I'm little bit doubtful it remove it, especially in cases of strange compilation errors.


Example single file:
```
[2024-04-10T12:22:10,615][DEBUG][org.logstash.config.ir.PipelineConfig] -------- Logstash Config ---------
[2024-04-10T12:22:10,615][DEBUG][org.logstash.config.ir.PipelineConfig] Config from source, source: LogStash::Config::Source::MultiLocal, pipeline_id:: test
[2024-04-10T12:22:10,615][DEBUG][org.logstash.config.ir.PipelineConfig] Config string, protocol: string, id: config_string
[2024-04-10T12:22:10,615][DEBUG][org.logstash.config.ir.PipelineConfig]

input { generator {} } #${I_AM_NOT_HERE}
filter { sleep { time => 100 } }
output {
#${I_AM_NOT_HERE}
  stdout {
    codec => rubydebug
  }
  # ${I_AM_NOT_HERE}
}

[2024-04-10T12:22:10,615][DEBUG][org.logstash.config.ir.PipelineConfig] Merged config
[2024-04-10T12:22:10,615][DEBUG][org.logstash.config.ir.PipelineConfig]

input { generator {} } #${I_AM_NOT_HERE}
filter { sleep { time => 100 } }
output {
#${I_AM_NOT_HERE}
  stdout {
    codec => rubydebug
  }
  # ${I_AM_NOT_HERE}
}

[2024-04-10T12:22:10,616][DEBUG][logstash.agent           ] Converging pipelines state {:actions_count=>1}
```

Example multiple files:
```
[2024-04-10T12:29:45,178][DEBUG][org.logstash.config.ir.PipelineConfig] -------- Logstash Config ---------
[2024-04-10T12:29:45,178][DEBUG][org.logstash.config.ir.PipelineConfig] Config from source, source: LogStash::Config::Source::MultiLocal, pipeline_id:: test
[2024-04-10T12:29:45,178][DEBUG][org.logstash.config.ir.PipelineConfig] Config string, protocol: file, id: /tmp/logstash_1.config
[2024-04-10T12:29:45,179][DEBUG][org.logstash.config.ir.PipelineConfig]

input { generator {} } #${I_AM_NOT_HERE}
filter { sleep { time => 100 } }
[2024-04-10T12:29:45,179][DEBUG][org.logstash.config.ir.PipelineConfig] Config string, protocol: file, id: /tmp/logstash_2.config
[2024-04-10T12:29:45,179][DEBUG][org.logstash.config.ir.PipelineConfig]

output {
     #${I_AM_NOT_HERE}
       stdout {
         codec => rubydebug
       }
       # ${I_AM_NOT_HERE}
}
[2024-04-10T12:29:45,179][DEBUG][org.logstash.config.ir.PipelineConfig] Merged config
[2024-04-10T12:29:45,179][DEBUG][org.logstash.config.ir.PipelineConfig]

input { generator {} } #${I_AM_NOT_HERE}
filter { sleep { time => 100 } }
output {
     #${I_AM_NOT_HERE}
       stdout {
         codec => rubydebug
       }
       # ${I_AM_NOT_HERE}
}
[2024-04-10T12:29:45,179][DEBUG][logstash.agent           ] Converging pipelines state {:actions_count=>1}
```

**Review by andsel — CHANGES_REQUESTED** [review]
Globally seems good, but left a doubt on the effectiveness of removing the debug line that prints the full pipeline string passed down to the compiler.

File: logstash-core/src/main/java/org/logstash/config/ir/PipelineConfig.java [hunk 7]
```diff
@@ -172,7 +172,6 @@ public void displayDebugInformation() {
             logger.debug("\n\n{}", configPart.getText());
         }
         logger.debug("Merged config");
```
Comment (mashhurs) [inline logstash-core/src/main/java/org/logstash/config/ir/PipelineConfig.java:174]:
```suggestion
        logger.debug("Merged config");
        logger.debug("\n\n{}", this.configString());
```

**Review by andsel — APPROVED** [review]
LGTM

**Pull request merged**